sql/src/main/java/io/crate/operation/projectors/BatchConsumerToRowReceiver.java

```diff
@@ -27,25 +27,55 @@
 import io.crate.data.Row;
 import io.crate.data.RowBridging;
 import io.crate.exceptions.SQLExceptions;
+import io.crate.data.Killable;
+import org.elasticsearch.common.collect.Tuple;
 
+import javax.annotation.Nullable;
 import java.util.Objects;
+import java.util.function.Function;
 
-public class BatchConsumerToRowReceiver implements BatchConsumer {
+public class BatchConsumerToRowReceiver implements BatchConsumer, Killable {
 
-    private final RowReceiver rowReceiver;
+    private RowReceiver rowReceiver;
+    private volatile boolean running = false;
 
     public BatchConsumerToRowReceiver(RowReceiver rowReceiver) {
         Objects.requireNonNull(rowReceiver, "RowReceiver cannot be null");
         this.rowReceiver = rowReceiver;
     }
 
+    private BatchIterator applyProjections(BatchIterator iterator) {
+        RowReceiver receiver = rowReceiver;
+        while (receiver instanceof Projector) {
+            Function<BatchIterator, Tuple<BatchIterator, RowReceiver>> projection = ((Projector) receiver).batchIteratorProjection();
+            if (projection == null) {
+                break;
+            }
+            Tuple<BatchIterator, RowReceiver> tuple = projection.apply(iterator);
+            iterator = tuple.v1();
+            receiver = tuple.v2();
+        }
+        rowReceiver = receiver;
+        return iterator;
+    }
+
     @Override
-    public void accept(BatchIterator iterator, Throwable failure) {
-        rowReceiver.completionFuture().whenComplete((ignored, t) -> iterator.close());
-        if (failure == null) {
-            safeConsumeIterator(iterator);
+    public void accept(BatchIterator it, Throwable failure) {
```

Comment:
you should handle failure here and not call applyProjections if failure is not null